src/main/java/org/kairosdb/core/datastore/QueryQueuingManager.java

```diff
@@ -113,6 +114,39 @@ private boolean acquireSemaphore(String queryHash, QueryMetric metric) throws In
 			return true;
 	}
 
+	public ArrayList<Pair<String, QueryMetric>> getRunningQueries()
+	{
+		ArrayList<Pair<String, QueryMetric>> runningQueriesList = new ArrayList<Pair<String, QueryMetric>>();
+		lock.lock();
+		ArrayList<String> queryKeys = new ArrayList<String>();
+		queryKeys.addAll(runningQueries.keySet());
+		try
+		{
+			for (String key : queryKeys)
+			{
+				runningQueriesList.add(new Pair<String, QueryMetric>(key, runningQueries.get(key).getFirst()));
+			}
+		}
+		finally
+		{
+			lock.unlock();
+		}
+		return runningQueriesList;
+	}
+
+	public void killQuery(String queryHash)
+	{
+		lock.lock();
+		try
+		{
+			runningQueries.get(queryHash).getSecond().interrupt();	// Call interrupt on Thread associated with provided query hash
```

Comment:
You need to check for null after the first get as the query may have finished before you call this to kill it.

Reply:
@brianhks I've added an if statement to check if the runningQueries.get() results in a null, but should I add an else branch and throw an exception or return a message saying, "couldn't kill this query, it was already gone"? If the query is already gone, it seems to me that the desired result is already in effect.

Reply:
I wouldn't add an else.  Like you said if the query is gone then mission accomplished.